byterun/u8tou16.c

```diff
@@ -0,0 +1,161 @@
+#include "caml/u8tou16.h"
+
+/* Copyright 2005 b8_bavard, INRIA, CML */
+/*
+    This file is part of mldonkey.
+
+    mldonkey is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    mldonkey is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with mldonkey; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+/* Stub code to interface with libiconv
+ *
+ * taken and modified from
+ *   mldonkey/src/utils/lib/charsetstubs.c
+ *
+ */
+
+/*
+  20080415 ygrek
+  Modified to use WideCharToMultiByte windows api
+*/
+
+#ifdef HAS_WINAPI_UTF16
+
+#define STRICT      /* Strict typing, please */
+#include <windows.h>
+#undef STRICT
+
+#define FALSE 0
+#define TRUE  1
+
+WCHAR* convert_to_utf16(const char* str,
+                        size_t len,
+                        UINT codepage)
+{
+    WCHAR* outp;
+    size_t outbuf_size;
+    int chars;
+
+    outbuf_size = len*2 + 8;
+
+    outp = malloc(outbuf_size + 2);
+    memset(outp,0,outbuf_size + 2);
+
+    if (0 == len)
+    {
+        return outp;
+    }
+
+    chars = MultiByteToWideChar(codepage, 0, str, len, outp, outbuf_size);
+    if (0 == chars)
+    {
+        free(outp);
+        return NULL;
+    }
+
+    *(outp + chars) = (WCHAR)0;
+    return outp;
+}
+
+
+char* convert_from_utf16(const WCHAR* str,
+                         size_t len,
+                         UINT codepage)
+{
+    char* outp;
+    size_t outbuf_size;
+    int chars;
+
+    outbuf_size = len*2 + 8;
+
+    outp = malloc(outbuf_size + 2);
```

Comment:
Indeed.  I guess the safe way is to call `WideCharToMultiByte` twice, the first one to get the length of the output buffer and the second one to actually convert the string.